### PR TITLE
Fix issue 253

### PR DESCRIPTION
**Pull request by Felalolf**
What I intended to solve with the `assert` function from the ViperWriter is more complicated than I had originally thought. We would need some kind of sophisticated dependency analysis to not violate any reordering constraints.
For now, I simplified the solution at the price of some performance. The assert function now always generates an additional function.

Additionally, I adapted the equality and string component such that they only generate a domain if necessary.

This PR only fixes issue 253. Somehow issue number 250, 251, 252 were skipped.

### CODE REVIEW

File: src/main/scala/viper/gobra/translator/encodings/StringEncoding.scala [hunk 1]
```diff
@@ -29,10 +29,15 @@ class StringEncoding extends LeafTypeEncoding {
     * Translates a type into a Viper type.
     */
   override def typ(ctx: Context): in.Type ==> vpr.Type = {
-    case ctx.String() / Exclusive => vpr.Int
+    case ctx.String() / Exclusive => stringType
     case ctx.String() / Shared => vpr.Ref
   }
 
+  private lazy val stringType: vpr.Type = {
```
Comment (jcp19) [inline src/main/scala/viper/gobra/translator/encodings/StringEncoding.scala:36]: Good call, I should have done this from the beginning

File: src/test/resources/regressions/issues/000253.gobra [hunk 2]
```diff
@@ -0,0 +1,15 @@
+// Any copyright is dedicated to the Public Domain.
+// http://creativecommons.org/publicdomain/zero/1.0/
+
+package issue250
```
Comment (jcp19) [inline src/test/resources/regressions/issues/000253.gobra:4]: nitpick: the package should be `issue253`

File: src/main/scala/viper/gobra/translator/util/ViperWriter.scala [hunk 3]
```diff
@@ -366,33 +339,54 @@ object ViperWriter {
...
 
+    /* Emits Viper statements. */
     def write(stmts: vpr.Stmt*): Writer[Unit] =
       create(stmts.toVector.map(Statement), ())
 
+    /* Emits Viper statements. */
     def assume(cond: vpr.Exp*): Writer[Unit] =
       create(cond.toVector.map(Assumption), ())
 
+    /* Can be used in expressions. */
```
Comment (jcp19) [inline src/main/scala/viper/gobra/translator/util/ViperWriter.scala:350]: :+1:

**Review by jcp19 — APPROVED** [review]

**Comment by ArquintL** [comment]
You do not see issues 250, 251, 252 as GitHub uses the same counter for issues and PRs

**Review by ArquintL — APPROVED** [review]

**Pull request merged**